androguard/core/apk/__init__.py

```diff
@@ -1501,33 +1503,200 @@ def is_androidtv(self) -> bool:
         """
         return self.get_attribute_value('uses-feature', 'name', required="false", name="android.hardware.touchscreen") == "android.hardware.touchscreen"
 
-    def get_certificate_der(self, filename:str) -> bytes:
+    def get_certificate_der(self, filename:str) -> Union[bytes, None]:
         """
         Return the DER coded X.509 certificate from the signature file.
+        If minSdkVersion is prior to Android N only the first SignerInfo is used.
+        If signed attributes are present, they are taken into account
 
         :param filename: Signature filename in APK
-        :returns: DER coded X.509 certificate as binary
+        :returns: DER coded X.509 certificate as binary or None
         """
+
+        # Get the signature
         pkcs7message = self.get_file(filename)
+        # Get the .SF
+        sf_filename = os.path.splitext(filename)[0] + '.SF'
+        sf_object = self.get_file(sf_filename)
+        # Load the signature
+        signed_data = cms.ContentInfo.load(pkcs7message)
+        # Locate the SignerInfo structure
+        signer_infos = signed_data['content']['signer_infos']
+        if not signer_infos:
+            raise ValueError('No signer information found in the PKCS7 object. The APK may not be properly signed.')
+
+        # Prior to Android N, Android attempts to verify only the first SignerInfo. From N onwards, Android attempts
+        # to verify all SignerInfos and then picks the first verified SignerInfo.
+        min_sdk_version = self.get_min_sdk_version()
+        if min_sdk_version is None or int(min_sdk_version) < 24:  # AndroidSdkVersion.N
+            logger.info(f"minSdkVersion: {min_sdk_version} is less than 24. Getting the first signerInfo only!")
+            unverified_signer_infos_to_try = [signer_infos[0]]
+        else:
+            unverified_signer_infos_to_try = signer_infos
+
+        # Extract certificates from the PKCS7 object
+        certificates = signed_data['content']['certificates']
+        return_certificate = None
+        list_certificates_verified = []
+        for signer_info in unverified_signer_infos_to_try:
+            try:
+                matching_certificate_verified = self.verify_signer_info_against_sig_file(
+                    signed_data,
+                    certificates,
+                    signer_info,
+                    sf_object)
+            except ValueError as e:
```

Comment:
I'm not sure whether this should also catch a possible `TypeError` or `OSError` from oscrypto functions being called (an invalid cert seems to raise a `ValueError` as well but I'm not sure that's always the case).

Also not sure if ASN.1 parsing errors should return `None` instead.